website/content/docs/secrets/openldap.mdx

```diff
@@ -172,11 +175,125 @@ only Vault will know.
+password and enable the account.
+
+- Passwords in AD are set using the `unicodePwd` field. This must be proceeded by two (2) colons (`::`).
+- When setting a password programatically in AD, the following critera must be met:
+  - The password must be enclosed in double quotes (`" "`)
+  - The password must be in [`UTF16LE` format](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/6e803168-f140-4d23-b2d3-c3a8ab5917d2)
+  - The password must be `base64`-encoded
+  - Additional details can be found [here](https://docs.microsoft.com/en-us/troubleshoot/windows-server/identity/set-user-password-with-ldifde)
+
+- Once a user's password has been set, it can be enabled. AD uses the `userAccountControl` field for this purpose:
+  - To enable the account, set `userAccountControl` to `512`
+  - You will likely also want to disable AD's password expiration for this dynamic user account. The
+`userAccountControl` value for this is: `65536`
+  - `userAccountControl` flags are cumulative, so to set both of the above two flags, add up the two values
+(`512 + 65536 = 66048`): set `userAccountControl` to `66048`
+  - See [here](https://docs.microsoft.com/en-us/troubleshoot/windows-server/identity/useraccountcontrol-manipulate-account-properties#property-flag-descriptions)
+for details on `userAccountControl` flags
+
+`sAMAccountName` is a common field when working with AD users. It is used to provide compatibility with legacy
+Windows NT systems and has a limit of 20 characters. Keep this in mind when defining your `username_template`.
+See [here](fhttps://docs.microsoft.com/en-us/windows/win32/adschema/a-samaccountname) for additional details.
+
+With regard to adding dynamic users to groups, AD doesn't let you directly modify a user's `memberOf` attribute.
+The `member` attribute of a group and `memberOf` attribute of a user are
+[linked attributes](https://docs.microsoft.com/en-us/windows/win32/ad/linked-attributes). Linked attributes are
+forward link/back link pairs, with the forward link able to be modified. In the case of AD group membership, the
+group's `member` attribute is the forward link. In order to add a newly-created dynamic user to a group, we also
+need to issue a `modify` request to the desired group and update the group membership with the new user.
+
+#### Active Directory LDIF Example
+
+The various `*_ldif` parameters are templates that use the [go template](https://golang.org/pkg/text/template/)
+language. A complete LDIF example for creating an Active Directory user account is provided here for reference:
+
+```ldif
+dn: CN={{.Username}},OU=HashiVault,DC=adtesting,DC=lab
+changetype: add
+objectClass: top
+objectClass: person
+objectClass: organizationalPerson
+objectClass: user
+userPrincipalName: {{.Username}}@adtesting.lab
+sAMAccountName: {{.Username}}
+
+dn: CN={{.Username}},OU=HashiVault,DC=adtesting,DC=lab
+changetype: modify
+replace: unicodePwd
+unicodePwd::{{ printf "\"%s\"" .Password | utf16le | base64 }}
```

Comment:
They can use `%q` to quote the string right?

```suggestion
unicodePwd::{{ printf "%q" .Password | utf16le | base64 }}
```

Reply:
`%q` returns single quoted values, not double quotes ones so that would be different =/

Reply:
> %q	a double-quoted string safely escaped with Go syntax

`%q` should return double-quoted strings: https://play.golang.org/p/3ZObvmbBMAM

Reply:
I stand corrected. The docs state `a single-quoted character literal safely escaped with Go syntax.` which looks like it's wrong.

Reply:
It looks like that's the case (single quoted) if the value is an integer, but it's double quotes for strings and byte slices.

Reply:
Integers don't appear to come out as quoted string representations of integers: https://play.golang.com/p/dhRW2zYyb1l
```
Raw value: [](string)
  => [""]
Raw value: [string](string)
  => ["string"]
Raw value: [77](int)
  => ['M']
Raw value: [77](int)
  => ['M']
Raw value: [3892](int)
  => ['༴']
Raw value: [77](float64)
  => [%!q(float64=77)]
Raw value: [true](bool)
  => [%!q(bool=true)]
```

Reply:
Isn't `.Password | utf16le | base64` always going to result in a string though? In this example `%q` should work fine.

Reply:
Yes it will return a string. I'm not disagreeing with that point. I was just pointing out that `%q` is not a "string quoting" verb.

Reply:
Gotcha. That's right, `%q` is not specific to string quoting and behaves differently depending on the value's type.